File: constructBTInPost.py
# Leetcode 106. Construct Binary Tree from Inorder and Postorder Traversal

# Time Complexity :  O(n) where n is the number of the nodes in tree

# Space Complexity : O(n) where n is the number of the nodes in tree thath can be in the stack in worst case

# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

# Approach: start at the root and traverse down the tree on the left child. Append the root to stack at
# every level of the tree. Once we reach the leaf, we check for its children and we exit out of the inner
# while i.e. we reached the left leaf. Now pop the stack and use the element as root. Check if there is a
# previous node and if its val is greater than root. If its greater than return false as that is not possible
# in a valid BST. Now traverse the right child. Once all the nodes are travsersed and stack is empty, return
# True

# Your code here along with comments explaining your approach

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    # global hashmap to store inorder elements and their indices
    mapp = dict()
    # Pointer to find root index from pre order
    index = None

    def buildTree(self, inorder: List[int], postorder: List[int]) -> TreeNode:
        self.index = len(postorder)-1
        # base
        if len(postorder) == 0 and len(inorder) == 0:
            return None
        # Filling the hashmap using inorder
        for i in range(len(inorder)):
            self.mapp[inorder[i]] = i
        # helper function call with start and ending indices for sub trees
        return self.helper(postorder, inorder, 0, len(inorder)-1)

    def helper(self, postorder, inorder, start, end):
        # base
        # When pointer on postorder reaches start of the array or start and end indices meet
        if self.index == -1 or start > end:
            return None

        # Logic
        # Using postorder and a pointer to find rootIndex at every step
        rootIndex = self.mapp[postorder[self.index]]
        # Using to rootIndex to find the root node from Inorder
        root = TreeNode(inorder[rootIndex])
        # Incrementing the index to find next root node index from postorder
        self.index -= 1
        # recurssive call for right sub tree
        root.right = self.helper(postorder, inorder, rootIndex+1, end)
        # recurssive call for left sub tree
        root.left = self.helper(postorder, inorder, start, rootIndex-1)

        return root

File: ConstructBTInPost.java
/*
 * #106. Construct Binary Tree from Inorder and Postorder Traversal
 * 
 * Given inorder and postorder traversal of a tree, construct the binary tree.

Note:
You may assume that duplicates do not exist in the tree.

For example, given

inorder = [9,3,15,20,7]
postorder = [9,15,7,20,3]

Return the following binary tree:

    3
   / \
  9  20
    /  \
   15   7
   
 */


/*
 * Time Complexity: O (N*N) -> O (N^2) // N - Traverse all nodes of a tree while constructing tree
 * 									   // N - For each node, traverse through N or N-1 or N-2 etc elements in a postorder array to find root for every recursive call
 * 
 * Space Complexity: O (N) + O (N*M) -> O (N*M) // N - Recursive stack space to store 'N' nodes of a tree
 * 							 		// (N*M) - total space for creating copy of arrays for each node of a tree in every recursive call
 * 
 * Did this code successfully run on leetcode: Yes
 * 
 * Any problem you faced while coding this: No
 * 
 */

package com.s30.edu.trees2;

import java.util.Arrays;

//Definition for a binary tree node.
class TreeNode {
	int val;
	TreeNode left;
	TreeNode right;
	TreeNode() {}
	TreeNode(int val) { this.val = val; }
	TreeNode(int val, TreeNode left, TreeNode right) {
		this.val = val;
		this.left = left;
		this.right = right;
	}
}


public class ConstructBTInPost {
public TreeNode buildTree(int[] inorder, int[] postorder) {
        
        // Base condition
        if(inorder.length == 0 || postorder.length == 0){
            return null; // Do nothing
        }
        
        // Find the root element from postorder array for every recursive call
        // Initially, we know the root is last element in postorder array
        TreeNode root = new TreeNode(postorder[postorder.length - 1]);
        
        // Find the index of root element in inorder array for every recursive call
        int index = -1; // Initialize the index
        
        // For 1 node, this will run O(N) times
        // For N nodes, this will run O(N*N) = O(N^2)
        // Time Complexity: O (N^2)
        for(int i = 0; i < inorder.length; i++){
            if(inorder[i] == root.val){
                index = i;
                break;
            }
        }
        
        // Find the left subtree size and right subtree size in inorder array for every recursive call
        int leftSize = index - 0;
        int rightSize = inorder.length - 1 - index; // We are just using leftSize in code
        
        // Create the 4 arrays as a copy of respective original array for every recursive call
        // inorder left, inorder right, postorder left and postorder right 
        
        // e.g: inLeft = (inorder, 0, 1) -> 1 is exclusive, last index is exclusive
        // We are dividing the given problems into smaller subproblems
        int[] inLeft = Arrays.copyOfRange(inorder, 0, index);
        int[] postLeft = Arrays.copyOfRange(postorder, 0, leftSize);
        
        int[] inRight = Arrays.copyOfRange(inorder, index+1, inorder.length);
        int[] postRight = Arrays.copyOfRange(postorder, leftSize, postorder.length-1);
        
        // Now, apply the recursion on both left and right arrays (for inorder and postorder both) to build a tree
        root.left = buildTree(inLeft, postLeft);
        root.right = buildTree(inRight, postRight);
        
        // return root (whole binary tree)
        return root;               
    }
    
}
